types/react/index.d.ts

```diff
@@ -90,6 +90,23 @@ declare namespace React {
     type RefCallback<T> = { bivarianceHack(instance: T | null): void }["bivarianceHack"];
     type Ref<T> = RefCallback<T> | RefObject<T> | null;
     type LegacyRef<T> = string | Ref<T>;
+    /**
+     * Gets the instance type for a React element. The instance will be different for various component types:
+     * 
+     * - React class components will be the class instance. So if you had `class Foo extends React.Component<{}> {}` and used `React.ElementRef<typeof Foo>` then the type would be the instance of `Foo`.
+     * - React stateless functional components do not have a backing instance and so `React.ElementRef<typeof Bar>` (when `Bar` is `function Bar() {}`) will give you the `undefined` type.
+     * - JSX intrinsics like `div` will give you their DOM instance. For `React.ElementRef<'div'>` that would be `HTMLDivElement`. For `React.ElementRef<'input'>` that would be `HTMLInputElement`.
+     * 
+     * `C` must be the type _of_ a React component so you need to use typeof as in React.ElementRef<typeof MyComponent>.
+     */
+    type ElementRef<C extends ComponentClass<any> | FunctionComponent<any> | keyof ElementTagNameMap> =
+        C extends React.ComponentClass<any>
+        ? InstanceType<C>
+        : C extends FunctionComponent<any>
+        ? undefined
+        : C extends keyof ElementTagNameMap
+        ? ElementTagNameMap[C]
```

Comment:
is there any other way you can think of to achieve this without adding the `ElementTagNameMap`?  It'd be really great to avoid adding this, due to operational overhead as well as risk of getting out of sync as new `HTMLXElement` types become necessary to add to other maps of this sort in the react types.  If nothing else, an `Omit` used on another map opting out the current types we don't want to be included here would be an improvement.  Let me know if you want a concrete example of what I'm talking about with the Omit but hopefully you catch my drift. :)

Reply:
Yup, I catch your drift 👍 At first I found only `ReactHTML` and `ReactSVG`, which would not have been sufficient, but I just did another pass and found that `JSX.IntrinsicElements` is defined in the react typings and I think that should suffice (looking now).

Reply:
Done, thanks for the nudge 🙏